plugins/hls-class-plugin/src/Ide/Plugin/Class/Types.hs

```diff
@@ -81,30 +115,101 @@ data BindInfo = BindInfo
+
+        getClsInstD (ClsInstD _ d) = Just d
+        getClsInstD _              = Nothing
+
+        getSigName (ClassOpSig _ _ sigNames _) = Just $ map unLoc sigNames
+        getSigName _                           = Nothing
+
+        getBindSpanWithoutSig :: ClsInstDecl GhcRn -> [BindInfo]
+        getBindSpanWithoutSig ClsInstDecl{..} =
+            let bindNames = mapMaybe go (bagToList cid_binds)
+                go (L l bind) = case bind of
+                    FunBind{..}
+                        -- `Generated` tagged for Template Haskell,
+                        -- here we filter out nonsense generated bindings
+                        -- that are nonsense for displaying code lenses.
+                        --
+                        -- See https://github.com/haskell/haskell-language-server/issues/3319
+                        | not $ isGenerated (groupOrigin fun_matches)
+                            -> Just $ L l fun_id
+                    _       -> Nothing
+                -- Existed signatures' name
+                sigNames = concat $ mapMaybe (\(L _ r) -> getSigName r) cid_sigs
+                toBindInfo (L l (L l' _)) = BindInfo
+                    (locA l) -- bindSpan
+                    (locA l') -- bindNameSpan
+            in toBindInfo <$> filter (\(L _ name) -> unLoc name `notElem` sigNames) bindNames
+        getBindSpanWithoutSig _ = []
+
+        -- Get bind definition range with its rendered signature text
+        getRangeWithSig :: InstanceBindTypeSig -> IO (Maybe (Range, Int, T.Text))
+        getRangeWithSig bind = runMaybeT $ do
+            span <- hoistMaybe $ bindDefSpan bind
+            range <- hoistMaybe $ srcSpanToRange span
+            uniqueID <- liftIO $ hashUnique <$> newUnique
+            pure (range, uniqueID, bindRendered bind)
+
+
+
+getInstanceBindTypeSigsRule :: Recorder (WithPriority Log) -> Rules ()
+getInstanceBindTypeSigsRule recorder = do
+    defineNoDiagnostics (cmapWithPrio LogShake recorder) $ \GetInstanceBindTypeSigs nfp -> runMaybeT $ do
+        (tmrTypechecked -> gblEnv ) <- useMT TypeCheck nfp
+        (hscEnv -> hsc) <- useMT GhcSession nfp
+        let binds = collectHsBindsBinders $ tcg_binds gblEnv
+        (_, maybe [] catMaybes -> instanceBinds) <- liftIO $
+            initTcWithGbl hsc gblEnv ghostSpan $ traverse (bindToSig hsc gblEnv) binds
+        pure $ InstanceBindTypeSigsResult instanceBinds
+    where
+        rdrEnv gblEnv= tcg_rdr_env gblEnv
+        showDoc hsc gblEnv ty = showSDocForUser' hsc (mkPrintUnqualifiedDefault hsc (rdrEnv gblEnv)) (pprSigmaType ty)
+        bindToSig hsc gblEnv id = do
+            let name = idName id
+            whenMaybe (isBindingName name) $ do
+                env <- tcInitTidyEnv
+                let (_, ty) = tidyOpenType env (idType id)
+                pure $ InstanceBindTypeSig name
+                        (prettyBindingNameString (printOutputable name) <> " :: " <> T.pack (showDoc hsc gblEnv ty))
```

Comment:
This field of the rule result is strict, which means we're probably doing all this formatting work up-front, which I suspect we don't in fact want? Indeed, I think that's some of the primary stuff we want to delay until the resolve happens.